cirq/experiments/fidelity_estimation.py

```diff
@@ -227,3 +227,120 @@ def log_xeb_fidelity(
+
+        f = (∑_U (m_U - u_U) * (e_U - u_U)) / (∑_U (e_U - u_U)^2).
+
+    Args:
+        measured_expectations: A list of the m_U, the experimental estimates
+            of the observable, one for each circuit U.
+        exact_expectations: A list of the e_U, the exact value of the
+            observable. The order should match the order of the
+            `measured_expectations` argument.
+        uniform_expectations: A list of the u_U, the expectation of the
+            observable on a uniformly random bitstring. The order should match
+            the order in the other arguments.
+    """
+    numerator = 0.0
+    denominator = 0.0
+    for m, e, u in zip(measured_expectations, exact_expectations,
+                       uniform_expectations):
+        numerator += (m - u) * (e - u)
+        denominator += (e - u)**2
+    return numerator / denominator
+
+
+def least_squares_xeb_fidelity_from_probabilities(
+        hilbert_space_dimension: int,
+        observed_probabilities: Sequence[Sequence[float]],
+        all_probabilities: Sequence[Sequence[float]],
+        observable_from_probability: Optional[Callable[[float], float]] = None,
+        normalize_probabilities: bool = True) -> float:
+    """Least squares fidelity estimator with observable based on probabilities.
+
+    Using the notation from the docstring of
+    `least_squares_xeb_fidelity_from_expectations`, this function computes the
+    least squares fidelity estimator when the observable O_U has eigenvalue
+    corresponding to the computational basis state |z⟩ given by g(p(z)), where
+    p(z) = |⟨z|𝜓_U⟩|^2 and g is a function that can be specified. By default,
+    g is the identity function, but other choices, such as the logarithm, are
+    useful. By default, the probability p(z) is actually multiplied by the
+    Hilbert space dimension D, so that the observable is actually g(D * p(z)).
+    This behavior can be disabled by setting `normalize_probabilities` to
+    False.
+
+    Args:
+        hilbert_space_dimension: Dimension of the Hilbert space on which
+           the channel whose fidelity is being estimated is defined.
+        observed_probabilities: Ideal probabilities of bitstrings observed in
+            experiments. A list of lists, where each inner list contains the
+            probabilities for a single circuit.
+        all_probabilities: Ideal probabilities of all possible bitstrings.
+            A list of lists, where each inner list contains the probabilities
+            for a single circuit, and should have length equal to the Hilbert
+            space dimension. The order of the lists should correspond to that
+            of `observed_probabilities`.
+        observable_from_probability: Function that computes the observable from
+            a given probability.
+        normalize_probabilities: Whether to multiply the probabilities by the
+            Hilbert space dimension before computing the observable.
```

Comment:
This argument has a very narrow use and complicates the API. We can make the API simpler and at the same time more flexible by removing `normalize_probabilities` and instead changing the expected signature of `observable_from_probability` from

```python
Callable[[float], float]
```

to

```python
Callable[[float, int], float]
```

where the second argument to `observable_from_probability` is the dimension of the Hilbert space. This means that identity which is currently specified as `lambda p: p` below becomes `lambda p, _: p`. Thus, user is free to ignore the second argument if their function is dimension-independent, but can also use it to normalize the first argument (probability) or to do something else (hence extra flexibility of this API).

Reply:
It's crucial for performance that `observable_from_probability` is allowed to be a built-in Numpy ufunc, since it is applied to Numpy arrays that are likely to have thousands to millions of elements. Your proposed change would preclude that.

Reply:
Makes sense.